File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OutcomeEventsCache.java
package com.onesignal;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.support.annotation.WorkerThread;

import com.onesignal.OneSignalDbContract.OutcomeEventsTable;
import com.onesignal.OneSignalDbContract.CachedUniqueOutcomeNotificationTable;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.List;

class OutcomeEventsCache {

    /**
     * Delete event from the DB
     */
    @WorkerThread
    synchronized static void deleteOldOutcomeEvent(OutcomeEvent event, OneSignalDbHelper dbHelper) {
        SQLiteDatabase writableDb = dbHelper.getWritableDbWithRetries();

        try {
            writableDb.beginTransaction();
            writableDb.delete(OutcomeEventsTable.TABLE_NAME,
                    OutcomeEventsTable.COLUMN_NAME_TIMESTAMP + " = ?", new String[]{String.valueOf(event.getTimestamp())});
            writableDb.setTransactionSuccessful();
        } catch (SQLiteException e) {
            OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Error deleting old outcome event records! ", e);
        } finally {
            if (writableDb != null) {
                try {
                    writableDb.endTransaction(); // May throw if transaction was never opened or DB is full.
                } catch (SQLiteException e) {
                    OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Error closing transaction! ", e);
                }
            }
        }
    }

    /**
     * Save an outcome event to send it on the future
     * <p>
     * For offline mode and contingency of errors
     */
    @WorkerThread
    synchronized static void saveOutcomeEvent(OutcomeEvent event, OneSignalDbHelper dbHelper) {
        SQLiteDatabase writableDb = dbHelper.getWritableDbWithRetries();
        String notificationIds = event.getNotificationIds() != null ? event.getNotificationIds().toString() : "[]";

        ContentValues values = new ContentValues();
        values.put(OutcomeEventsTable.COLUMN_NAME_NOTIFICATION_IDS, notificationIds);
        values.put(OutcomeEventsTable.COLUMN_NAME_SESSION, event.getSession().toString().toLowerCase());
        values.put(OutcomeEventsTable.COLUMN_NAME_NAME, event.getName());
        values.put(OutcomeEventsTable.COLUMN_NAME_TIMESTAMP, event.getTimestamp());
        values.put(OutcomeEventsTable.COLUMN_NAME_WEIGHT, event.getWeight());

        writableDb.insert(OutcomeEventsTable.TABLE_NAME, null, values);
    }

    /**
     * Save an outcome event to send it on the future
     * <p>
     * For offline mode and contingency of errors
     */
    @WorkerThread
    synchronized static List<OutcomeEvent> getAllEventsToSend(OneSignalDbHelper dbHelper) {
        List<OutcomeEvent> events = new ArrayList<>();
        Cursor cursor = null;

        try {
            SQLiteDatabase readableDb = dbHelper.getReadableDbWithRetries();
            cursor = readableDb.query(
                    OutcomeEventsTable.TABLE_NAME,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null
            );

            if (cursor.moveToFirst()) {
                do {
                    String sessionString = cursor.getString(cursor.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_SESSION));
                    OSSessionManager.Session session = OSSessionManager.Session.fromString(sessionString);
                    String notificationIds = cursor.getString(cursor.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_NOTIFICATION_IDS));
                    String name = cursor.getString(cursor.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_NAME));
                    long timestamp = cursor.getLong(cursor.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_TIMESTAMP));
                    float weight = cursor.getFloat(cursor.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_WEIGHT));

                    try {
                        OutcomeEvent event = new OutcomeEvent(session, new JSONArray(notificationIds), name, timestamp, weight);
                        events.add(event);

                    } catch (JSONException e) {
                        OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Generating JSONArray from notifications ids outcome:JSON Failed.", e);
                    }
                } while (cursor.moveToNext());
            }
        } finally {
            if (cursor != null && !cursor.isClosed())
                cursor.close();
        }

        return events;
    }

    /**
     * Save a JSONArray of notification ids as separate items with the unique outcome name
     */
    @WorkerThread
    synchronized static void saveUniqueOutcomeNotifications(JSONArray notificationIds, String outcomeName, OneSignalDbHelper dbHelper) {
        if (notificationIds == null)
            return;

        SQLiteDatabase writableDb = dbHelper.getWritableDbWithRetries();
        try {
            for (int i = 0; i < notificationIds.length(); i++) {
                ContentValues values = new ContentValues();

                String notificationId = notificationIds.getString(i);
                values.put(CachedUniqueOutcomeNotificationTable.COLUMN_NAME_NOTIFICATION_ID, notificationId);
                values.put(CachedUniqueOutcomeNotificationTable.COLUMN_NAME_NAME, outcomeName);

                writableDb.insert(CachedUniqueOutcomeNotificationTable.TABLE_NAME, null, values);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    /**
     * Create a JSONArray of not cached notification ids from the unique outcome notifications SQL table
     */
    @WorkerThread
    synchronized static JSONArray getNotCachedUniqueOutcomeNotifications(String name, JSONArray notificationIds, OneSignalDbHelper dbHelper) {
        JSONArray uniqueNotificationIds = new JSONArray();

        SQLiteDatabase readableDb = dbHelper.getReadableDbWithRetries();
        Cursor cursor = null;

        try {
            for (int i = 0; i < notificationIds.length(); i++) {
                String notificationId = notificationIds.getString(i);
                CachedUniqueOutcomeNotification notification = new CachedUniqueOutcomeNotification(notificationId, name);

                String[] columns = new String[]{};

                String where = CachedUniqueOutcomeNotificationTable.COLUMN_NAME_NOTIFICATION_ID +  " = ? AND " +
                        CachedUniqueOutcomeNotificationTable.COLUMN_NAME_NAME +  " = ?";

                String[] args = new String[]{notification.getNotificationId(), notification.getName()};

                cursor = readableDb.query(
                        OneSignalDbContract.CachedUniqueOutcomeNotificationTable.TABLE_NAME,
                        columns,
                        where,
                        args,
                        null,
                        null,
                        null,
                        "1"
                );

                // Item is not cached, add it to the JSONArray
                if (cursor.getCount() == 0)
                    uniqueNotificationIds.put(notificationId);

            }
        } catch (JSONException e) {
            e.printStackTrace();
        } finally {
            if (cursor != null && !cursor.isClosed())
                cursor.close();
        }

        return uniqueNotificationIds;
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/core/internal/outcomes/impl/OutcomeEventsCache.kt
package com.onesignal.core.internal.outcomes.impl

import android.content.ContentValues
import com.onesignal.core.internal.database.IDatabaseProvider
import com.onesignal.core.internal.database.impl.OneSignalDbContract
import com.onesignal.core.internal.influence.Influence
import com.onesignal.core.internal.influence.InfluenceChannel
import com.onesignal.core.internal.influence.InfluenceType
import com.onesignal.core.internal.influence.InfluenceType.Companion.fromString
import com.onesignal.core.internal.logging.Logging
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONException
import java.util.Locale

internal class OutcomeEventsCache(
    private val dbHelper: IDatabaseProvider
) : IOutcomeEventsCache {

    /**
     * Delete event from the DB
     */
    override suspend fun deleteOldOutcomeEvent(event: OutcomeEventParams) {
        withContext(Dispatchers.IO) {
            dbHelper.get().delete(
                OutcomeEventsTable.TABLE_NAME,
                OutcomeEventsTable.COLUMN_NAME_TIMESTAMP + " = ?",
                arrayOf(event.timestamp.toString())
            )
        }
    }

    /**
     * Save an outcome event to send it on the future
     * For offline mode and contingency of errors
     */
    override suspend fun saveOutcomeEvent(eventParams: OutcomeEventParams) {
        withContext(Dispatchers.IO) {
            var notificationIds = JSONArray()
            var iamIds = JSONArray()
            var notificationInfluenceType = InfluenceType.UNATTRIBUTED
            var iamInfluenceType = InfluenceType.UNATTRIBUTED
            // Check for direct channels
            eventParams.outcomeSource?.directBody?.let { directBody ->
                directBody.notificationIds?.let {
                    if (it.length() > 0) {
                        notificationInfluenceType = InfluenceType.DIRECT
                        notificationIds = it
                    }
                }
                directBody.inAppMessagesIds?.let {
                    if (it.length() > 0) {
                        iamInfluenceType = InfluenceType.DIRECT
                        iamIds = it
                    }
                }
            }
            // Check for indirect channels
            eventParams.outcomeSource?.indirectBody?.let { indirectBody ->
                indirectBody.notificationIds?.let {
                    if (it.length() > 0) {
                        notificationInfluenceType = InfluenceType.INDIRECT
                        notificationIds = it
                    }
                }
                indirectBody.inAppMessagesIds?.let {
                    if (it.length() > 0) {
                        iamInfluenceType = InfluenceType.INDIRECT
                        iamIds = it
                    }
                }
            }
            ContentValues().apply {
                // Save influence ids
                put(OutcomeEventsTable.COLUMN_NAME_NOTIFICATION_IDS, notificationIds.toString())
                put(OutcomeEventsTable.COLUMN_NAME_IAM_IDS, iamIds.toString())
                // Save influence types
                put(OutcomeEventsTable.COLUMN_NAME_NOTIFICATION_INFLUENCE_TYPE, notificationInfluenceType.toString().toLowerCase())
                put(OutcomeEventsTable.COLUMN_NAME_IAM_INFLUENCE_TYPE, iamInfluenceType.toString().toLowerCase())
                // Save outcome data
                put(OutcomeEventsTable.COLUMN_NAME_NAME, eventParams.outcomeId)
                put(OutcomeEventsTable.COLUMN_NAME_WEIGHT, eventParams.weight)
                put(OutcomeEventsTable.COLUMN_NAME_TIMESTAMP, eventParams.timestamp)
            }.also { values ->
                dbHelper.get().insert(OutcomeEventsTable.TABLE_NAME, null, values)
            }
        }
    }

    /**
     * Save an outcome event to send it on the future
     * For offline mode and contingency of errors
     */
    override suspend fun getAllEventsToSend(): List<OutcomeEventParams> {
        val events: MutableList<OutcomeEventParams> = ArrayList()
        withContext(Dispatchers.IO) {
            dbHelper.get().query(
                OutcomeEventsTable.TABLE_NAME,
                null,
                null,
                null,
                null,
                null,
                null
            ).use {
                if (it.moveToFirst()) {
                    do {
                        // Retrieve influence types
                        val notificationInfluenceTypeString =
                            it.getString(it.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_NOTIFICATION_INFLUENCE_TYPE))
                        val notificationInfluenceType =
                            fromString(notificationInfluenceTypeString)
                        val iamInfluenceTypeString =
                            it.getString(it.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_IAM_INFLUENCE_TYPE))
                        val iamInfluenceType = fromString(iamInfluenceTypeString)

                        // Retrieve influence ids
                        val notificationIds =
                            it.getString(it.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_NOTIFICATION_IDS))
                                ?: "[]"
                        val iamIds =
                            it.getString(it.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_IAM_IDS))
                                ?: "[]"

                        // Retrieve Outcome data
                        val name =
                            it.getString(it.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_NAME))
                        val weight =
                            it.getFloat(it.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_WEIGHT))
                        val timestamp =
                            it.getLong(it.getColumnIndex(OutcomeEventsTable.COLUMN_NAME_TIMESTAMP))

                        try {
                            val directSourceBody = OutcomeSourceBody()
                            val indirectSourceBody = OutcomeSourceBody()
                            val source: OutcomeSource = getNotificationInfluenceSource(
                                notificationInfluenceType,
                                directSourceBody,
                                indirectSourceBody,
                                notificationIds
                            )
                                .also {
                                    getIAMInfluenceSource(
                                        iamInfluenceType,
                                        directSourceBody,
                                        indirectSourceBody,
                                        iamIds,
                                        it
                                    )
                                } ?: OutcomeSource(null, null)
                            OutcomeEventParams(name, source, weight, timestamp).also {
                                events.add(it)
                            }
                        } catch (e: JSONException) {
                            Logging.error(
                                "Generating JSONArray from notifications ids outcome:JSON Failed.",
                                e
                            )
                        }
                    } while (it.moveToNext())
                }
            }
        }
        return events
    }

    private fun getNotificationInfluenceSource(
        notificationInfluenceType: InfluenceType,
        directSourceBody: OutcomeSourceBody,
        indirectSourceBody: OutcomeSourceBody,
        notificationIds: String
    ): OutcomeSource? {
        return when (notificationInfluenceType) {
            InfluenceType.DIRECT -> {
                directSourceBody.notificationIds = JSONArray(notificationIds)
                OutcomeSource(directSourceBody, null)
            }
            InfluenceType.INDIRECT -> {
                indirectSourceBody.notificationIds = JSONArray(notificationIds)
                OutcomeSource(null, indirectSourceBody)
            }
            else -> {
                null
            }
        }
    }

    private fun getIAMInfluenceSource(
        iamInfluenceType: InfluenceType,
        directSourceBody: OutcomeSourceBody,
        indirectSourceBody: OutcomeSourceBody,
        iamIds: String,
        source: OutcomeSource?
    ): OutcomeSource? {
        return when (iamInfluenceType) {
            InfluenceType.DIRECT -> {
                directSourceBody.inAppMessagesIds = JSONArray(iamIds)
                source?.setDirectBody(directSourceBody)
                    ?: OutcomeSource(directSourceBody, null)
            }
            InfluenceType.INDIRECT -> {
                indirectSourceBody.inAppMessagesIds = JSONArray(iamIds)
                source?.setIndirectBody(indirectSourceBody)
                    ?: OutcomeSource(null, indirectSourceBody)
            }
            else -> {
                source
            }
        }
    }

    private fun addIdToListFromChannel(cachedUniqueOutcomes: MutableList<CachedUniqueOutcome>, channelIds: JSONArray?, channel: InfluenceChannel) {
        channelIds?.let {
            for (i in 0 until it.length()) {
                try {
                    val influenceId = it.getString(i)
                    cachedUniqueOutcomes.add(CachedUniqueOutcome(influenceId, channel))
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        }
    }

    private fun addIdsToListFromSource(cachedUniqueOutcomes: MutableList<CachedUniqueOutcome>, sourceBody: OutcomeSourceBody?) {
        sourceBody?.let {
            val iamIds = it.inAppMessagesIds
            val notificationIds = it.notificationIds
            addIdToListFromChannel(cachedUniqueOutcomes, iamIds, InfluenceChannel.IAM)
            addIdToListFromChannel(cachedUniqueOutcomes, notificationIds, InfluenceChannel.NOTIFICATION)
        }
    }

    /**
     * Save a JSONArray of notification ids as separate items with the unique outcome name
     */
    override suspend fun saveUniqueOutcomeEventParams(eventParams: OutcomeEventParams) {
        Logging.debug("OutcomeEventsCache.saveUniqueOutcomeEventParams(eventParams: $eventParams)")
        val outcomeName = eventParams.outcomeId
        val cachedUniqueOutcomes: MutableList<CachedUniqueOutcome> = ArrayList()
        val directBody = eventParams.outcomeSource?.directBody
        val indirectBody = eventParams.outcomeSource?.indirectBody
        addIdsToListFromSource(cachedUniqueOutcomes, directBody)
        addIdsToListFromSource(cachedUniqueOutcomes, indirectBody)

        for (uniqueOutcome in cachedUniqueOutcomes) {
            ContentValues().apply {
                put(CachedUniqueOutcomeTable.COLUMN_CHANNEL_INFLUENCE_ID, uniqueOutcome.influenceId)
                put(CachedUniqueOutcomeTable.COLUMN_CHANNEL_TYPE, uniqueOutcome.channel.toString())
                put(CachedUniqueOutcomeTable.COLUMN_NAME_NAME, outcomeName)
            }.also { values ->
                dbHelper.get().insert(CachedUniqueOutcomeTable.TABLE_NAME, null, values)
            }
        }
    }

    /**
     * Create a JSONArray of not cached notification ids from the unique outcome notifications SQL table
     */
    override suspend fun getNotCachedUniqueInfluencesForOutcome(name: String, influences: List<Influence>): List<Influence> {
        val uniqueInfluences: MutableList<Influence> = ArrayList()

        try {
            for (influence in influences) {
                val availableInfluenceIds = JSONArray()
                val influenceIds = influence.ids ?: continue

                for (i in 0 until influenceIds.length()) {
                    val channelInfluenceId = influenceIds.getString(i)
                    val channel = influence.influenceChannel
                    val columns = arrayOf<String>()
                    val where = CachedUniqueOutcomeTable.COLUMN_CHANNEL_INFLUENCE_ID + " = ? AND " +
                        CachedUniqueOutcomeTable.COLUMN_CHANNEL_TYPE + " = ? AND " +
                        CachedUniqueOutcomeTable.COLUMN_NAME_NAME + " = ?"
                    val args = arrayOf(channelInfluenceId, channel.toString(), name)
                    dbHelper.get().query(
                        CachedUniqueOutcomeTable.TABLE_NAME,
                        columns,
                        where,
                        args,
                        null,
                        null,
                        null,
                        "1"
                    ).use {
                        // Item is not cached, we can use the influence id, add it to the JSONArray
                        if (it.count == 0) availableInfluenceIds.put(channelInfluenceId)
                    }
                }

                if (availableInfluenceIds.length() > 0) {
                    influence.copy().apply {
                        ids = availableInfluenceIds
                    }.also { newInfluence ->
                        uniqueInfluences.add(newInfluence)
                    }
                }
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return uniqueInfluences
    }

    /**
     * Deletes cached unique outcome notifications whose ids do not exist inside of the NotificationTable.TABLE_NAME
     */
    override suspend fun cleanCachedUniqueOutcomeEventNotifications() {
        val notificationTableName = OneSignalDbContract.NotificationTable.TABLE_NAME
        val notificationIdColumnName = OneSignalDbContract.NotificationTable.COLUMN_NAME_NOTIFICATION_ID

        withContext(Dispatchers.IO) {
            val whereStr = "NOT EXISTS(" +
                "SELECT NULL FROM " + notificationTableName + " n " +
                "WHERE" + " n." + notificationIdColumnName + " = " + OutcomesDbContract.CACHE_UNIQUE_OUTCOME_COLUMN_CHANNEL_INFLUENCE_ID +
                " AND " + OutcomesDbContract.CACHE_UNIQUE_OUTCOME_COLUMN_CHANNEL_TYPE + " = \"" + InfluenceChannel.NOTIFICATION.toString()
                .toLowerCase(Locale.ROOT) +
                "\")"
            dbHelper.get().delete(
                OutcomesDbContract.CACHE_UNIQUE_OUTCOME_TABLE,
                whereStr,
                null
            )
        }
    }
}
